internal/charmstore/store.go

```diff
@@ -133,17 +136,26 @@ func (s *Store) putArchive(archive blobstore.ReadSeekCloser) (blobName, blobHash
 // also adds the charm archive to the blob store.
 // This method is provided principally so that
 // tests can easily create content in the store.
-func (s *Store) AddCharmWithArchive(url *charm.Reference, ch charm.Charm) error {
+//
+// If purl is not nil then the charm will also be
+// available at the promulgated url specified.
+func (s *Store) AddCharmWithArchive(url, purl *charm.Reference, ch charm.Charm) error {
```

Comment:
And at this point we are 100% sure that promulgated in base entity is set to 1, right?